### PR TITLE
Dup sql before sending to instrumentation.

### PR DESCRIPTION
**Pull request by arthurnn**
We should not allow a subscriber to mutate the query to be executed.

I found this because we had some bad code that were doing a `gsub!` in place for check some things on the query, however the query was been mutated and been pushed as a wrong key to the cache.

thoughts @rafaelfranca @matthewd @sgrif 

Not sure if I should write a regression test somewhere else too? maybe the subscriber?

### CODE REVIEW

**Comment by rafaelfranca** [comment]
I'm not sure about this. Is responsibility of Rails to avoid people to write wrong code? Why only this and not `name`, `statement_name` and `binds`?

It does make sense to me, but we don't defensively protect any other place and I don't think we should not.

**Comment by sgrif** [comment]
I disagree, I'm :+1: on this change.

> I'm not sure about this. Is responsibility of Rails to avoid people to write wrong code? 

No, but we do have a contract that we don't expect the user to mutate this argument. We should make that explicit when possible.

> we don't defensively protect any other place and I don't think we should not.

Maybe we should be defensive in more places. ;)

**Comment by arthurnn** [comment]
> Why only this and not name, statement_name and binds?

Really fair question. I was wondering about that while writing the commit message, but wanted to see thoughts in a small changeset first.

This is indeed preventing users to make mistake, I made this once, and it was hard to track, so I felt Rails could easily fix this. 

I think we should freeze the `binds` too, as we use that. However `name` and `statement_name` are just log strings as far as I can tell, so we don't use them afterwards, even if the user mutates them that would not change anything I guess.

**Comment by arthurnn** [comment]
This gets even more confusing when changing adapters. 
For instance, on mysql2 adapter, if you mutate the query when it was a cache lookup you will get this error:

```
  1) Error:
QueryCacheTest#test_query_cache_freezes_they_key:
TypeError: can't dup NilClass
    /Users/arthurnn/src/rails/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb:85:in `dup'
    /Users/arthurnn/src/rails/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb:85:in `cache_sql'
    /Users/arthurnn/src/rails/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb:68:in `select_all'
    /Users/arthurnn/src/rails/activerecord/lib/active_record/querying.rb:39:in `find_by_sql'
    /Users/arthurnn/src/rails/activerecord/lib/active_record/statement_cache.rb:109:in `execute'
    /Users/arthurnn/src/rails/activerecord/lib/active_record/core.rb:163:in `find'
    activerecord/test/cases/query_cache_test.rb:260:in `block (3 levels) in test_query_cache_freezes_they_key'
    /Users/arthurnn/src/rails/activerecord/test/cases/test_case.rb:38:in `assert_queries'
    activerecord/test/cases/query_cache_test.rb:260:in `block (2 levels) in test_query_cache_freezes_they_key'
    /Users/arthurnn/src/rails/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb:32:in `cache'
    /Users/arthurnn/src/rails/activerecord/lib/active_record/query_cache.rb:9:in `cache'
    activerecord/test/cases/query_cache_test.rb:259:in `block in test_query_cache_freezes_they_key'
    /Users/arthurnn/src/rails/activesupport/lib/active_support/notifications.rb:176:in `subscribed'
    activerecord/test/cases/query_cache_test.rb:258:in `test_query_cache_freezes_they_key'
```
I guess is because this [line](https://github.com/rails/rails/blob/fcec068bf913d74146703a02388608aa54d4f3a8/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb#L81) wont return anything as the `sql` var was mutated. 

Anyways, I think we are safer to not allowing mutations on those two variables as we use them. 

Another solution would be to `.dup` them before sending to instrumentation. Thoughts?

File: activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb [hunk 1]
```diff
@@ -77,7 +77,7 @@ def cache_sql(sql, binds)
         result =
           if @query_cache[sql].key?(binds)
             ActiveSupport::Notifications.instrument("sql.active_record",
-              :sql => sql, :binds => binds, :name => "CACHE", :connection_id => object_id)
+              :sql => sql.freeze, :binds => binds.freeze, :name => "CACHE", :connection_id => object_id)
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb]: Can/should the caller do this freezing? Arguably, this is another manifestation of the same problem you were originally tracking: code mutating a value it received as a parameter, when it should really be leaving it alone.

The query cache lookups seem like they would benefit from using frozen values, too.

Comment (arthurnn) [inline activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb]: Yeah, mutating the values on the query cache is another problem for sure. I didn't want to fix it here, tho, as that part I am not sure how to implement yet.

**Comment by jeremy** [comment]
Sympathetic to the impulse to disallow this, but the freezing as a side effect of logging is a step back. Got an alternative? Freeze upstream? Or dup + freeze when logging and eat the hotspot allocation cost?

**Comment by rafaelfranca** [comment]
> Maybe we should be defensive in more places. ;)

Please don't! :pray: Defensive programming is source of a lot of crap.

**Comment by arthurnn** [comment]
Changed this, so, it wont freeze, and instead it will .dup the sql and binds in the source place only.

Freeze was not the best idea, as we should freeze the string once we create it. However we create that sql in multiple points on different adapters, so it would be a more scatter implementation.
Instead we can `.dup` the string before sending to the instrumentation.

@jeremy @sgrif @matthewd thoughts?

**Comment by sgrif** [comment]
While it's probably a low enough cost not to matter, I don't like the idea of introducing anything that has any amount of a performance hit for this.

**Comment by arthurnn** [comment]
> While it's probably a low enough cost not to matter, I don't like the idea of introducing anything that has any amount of a performance hit for this.

The drawback about the `dup` solution over the `freeze` one is the extra allocations. I dont think they are a big deal, as these are not hotspots. However I do understand your concern.

Not sure what is the best solution for this then. Maybe try to freeze up in the stack?

**Pull request closed without merging**